### PR TITLE
Add design requirements for QUIC packet demuxer

### PR DESCRIPTION
**Pull request by t8m**
These are initial requirements for the QUIC packet demuxer module.

### CODE REVIEW

File: doc/designs/quic-design/demuxer.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+Packet Demuxer
+==============
+
+This is a QUIC specific module that parses headers of incoming packets and
+decides what to do next.
+
+Demuxer requirements for MVP
+----------------------------
+
+These are requirements that were identified for MVP:
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:10]: These are _the_ requirements

File: doc/designs/quic-design/demuxer.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+- multiple QUIC packets in an UDP packet handling as packet coalescing
+  must be supported
+- client must discard packets that do not match existing connection ID
+- client must discard packets with version different from the one initially
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:15]: must discard _any_ packets with _a_ version different

File: doc/designs/quic-design/demuxer.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  selected
+
+Optional demuxer requirements
+-----------------------------
+
+These are optional features of client side demuxer, not required for MVP
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:21]: of _the_ client side

File: doc/designs/quic-design/demuxer.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+but otherwise desirable:
+
+- optionally trigger sending stateless reset packets if a received packet
+  on client is well-formed but does not belong to a known connection
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:25]: on _the_ client _side_ is

File: doc/designs/quic-design/demuxer.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+Demuxer requirements for server
+-------------------------------
+
+Further requirements after MVP for server support:
+
+- on servers packets can create a new connection potentially
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:32]: s/on servers/on the server side,/

File: doc/designs/quic-design/demuxer.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+- server side packet handling for unsupported version packets:
+  - trigger sending version negotiation packets if server receives a packet
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:34]: if _the_ server receives

File: doc/designs/quic-design/demuxer.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    with unsupported version and is large enough to initiate a new connection;
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:35]: with _an_ unsupported version and _which_ is large enough

File: doc/designs/quic-design/demuxer.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    limit the number of such packets with the same destination
+  - discard smaller packets with unsupported version
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:37]: unsupported version_s_

File: doc/designs/quic-design/demuxer.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+- packet handling on server for well-formed packets with supported version
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:38]: on server **_side_** for well-formed packets with supported version_**s**_

File: doc/designs/quic-design/demuxer.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  but with unknown connection IDs:
+  - if the packet is well-formed Initial packet, trigger creating a new
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:40]: is _a_ well-formed

trigger _the creation of_ a new

File: doc/designs/quic-design/demuxer.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    connection
+  - if the the packet is well-formed 0RTT packet, mark the packet to be
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:42]: if **the** packet is **a** well-formed

File: doc/designs/quic-design/demuxer.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    buffered for short time (as Initial packet might arrive late)
```
Comment (paulidale) [inline doc/designs/quic-design/demuxer.md:43]: for _a_ short _period of_ time (as _an_ Initial packet might arrive late_r_)

**Review by paulidale — COMMENTED** [review]
Grammar nits, looks good otherwise.

**Comment by t8m** [comment]
Fixup pushed addressing the grammar nits.

**Review by paulidale — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Ping @openssl/contractors for second review

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master branch. Thank you for the reviews.

**Pull request closed without merging**